Add snapshot IDs to PFB export

The Terra team needs the Snapshot ID added to the PFB export.  Please refer to this [Google Doc](https://docs.google.com/document/d/1yiHmUuBKqd6Td3aZAqUhtr5LWUwoBtNT65Uw2JsJjAk/edit?usp=sharing).

Estimate 4.

The below patch passes all unit tests, except for one failure in IT.

```
======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [catalog_complete] (catalog='anvil-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/daniel/repo/azul1/test/integration_test.py", line 377, in subTest
    yield
  File "/Users/daniel/repo/azul1/test/integration_test.py", line 1352, in _assert_catalog_complete
    self.assertSetEqual(indexed_fqids, expected_fqids)
AssertionError: Items in the second set but not the first:
TDRAnvilBundleFQID(uuid='748f9698-8d05-a0e8-85de-75c86e959873', version='2022-06-01T00:00:00.000000Z', source=TDRSourceRef(id='5f55901c-4bfc-4d33-b85d-bd9f812f57cd', spec=TDRSourceSpec(prefix=Prefix(common='', partition=1), type=<Type.bigquery: 'bigquery'>, domain=<Domain.gcp: 'gcp'>, subdomain='datarepo-dev-42c70e6a', name='ANVIL_CCDG_Sample_1_20230228_ANV5_202302281520')), table_name='anvil_dataset', batch_prefix=None)

----------------------------------------------------------------------
```


```diff
Index: test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json b/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json
--- a/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json   (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json   (date 1756832717256)
@@ -433,6 +433,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS"
         },
         "relations": []
Index: test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json b/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json
--- a/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json (date 1756832717260)
@@ -582,6 +582,11 @@
                             }
                         },
                         {
+                            "name": "source_datarepo_snapshot_id",
+                            "namespace": "anvil_dataset",
+                            "type": "string"
+                        },
+                        {
                             "name": "title",
                             "namespace": "anvil_dataset",
                             "type": [
Index: src/azul/plugins/repository/tdr_anvil/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr_anvil/__init__.py b/src/azul/plugins/repository/tdr_anvil/__init__.py
--- a/src/azul/plugins/repository/tdr_anvil/__init__.py (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/src/azul/plugins/repository/tdr_anvil/__init__.py (date 1756832717233)
@@ -473,6 +473,9 @@
         for entity_type, typed_keys in sorted(keys_by_type.items()):
             pk_column = entity_type.removeprefix('anvil_') + '_id'
             rows = self._retrieve_entities(source.spec, entity_type, typed_keys)
+            if entity_type == 'anvil_dataset':
+                for row in rows:
+                    row['source_datarepo_snapshot_id'] = source.id
             if entity_type == 'anvil_donor':
                 # We expect that the foreign key `part_of_dataset_id` is
                 # redundant for biosamples and donors. To simplify our queries,
@@ -495,7 +498,7 @@
 
     def _supplementary_bundle(self, bundle_fqid: TDRAnvilBundleFQID) -> TDRAnvilBundle:
         assert bundle_fqid.is_batched, bundle_fqid
-        source = bundle_fqid.source.spec
+        source = bundle_fqid.source
         result = TDRAnvilBundle(fqid=bundle_fqid)
         linked_file_refs = set()
         for file_ref, file_row in self._get_bundle_batch(bundle_fqid):
@@ -521,7 +524,7 @@
 
     def _duos_bundle(self, bundle_fqid: TDRAnvilBundleFQID) -> TDRAnvilBundle:
         assert not bundle_fqid.is_batched, bundle_fqid
-        ref, row = self._get_dataset(bundle_fqid.source.spec)
+        ref, row = self._get_dataset(bundle_fqid.source)
         expected_entity_id = change_version(bundle_fqid.uuid,
                                             self.bundle_uuid_version,
                                             self.datarepo_row_uuid_version)
@@ -541,7 +544,7 @@
 
     def _replica_bundle(self, bundle_fqid: TDRAnvilBundleFQID) -> TDRAnvilBundle:
         assert bundle_fqid.is_batched, bundle_fqid
-        source = bundle_fqid.source.spec
+        source = bundle_fqid.source
         result = TDRAnvilBundle(fqid=bundle_fqid)
         batch = self._get_bundle_batch(bundle_fqid)
         dataset = self._get_dataset(source)
@@ -549,13 +552,15 @@
             result.add_entity(ref, self._version, dict(row), is_orphan=True)
         return result
 
-    def _get_dataset(self, source: TDRSourceSpec) -> tuple[EntityReference, BigQueryRow]:
+    def _get_dataset(self, source: TDRSourceRef) -> tuple[EntityReference, BigQueryRow]:
         table_name = 'anvil_dataset'
         columns = self._columns(table_name)
         row = one(self._run_sql(f'''
             SELECT {', '.join(sorted(columns))}
-            FROM {backtick(self._full_table_name(source, table_name))}
+            FROM {backtick(self._full_table_name(source.spec, table_name))}
         '''))
+        row = dict(row, source_datarepo_snapshot_id=source.id)
+        # row['source_datarepo_snapshot_id'] = source.id
         ref = EntityReference(entity_type=table_name, entity_id=row['datarepo_row_id'])
         return ref, row
 
Index: src/azul/plugins/metadata/anvil/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/metadata/anvil/__init__.py b/src/azul/plugins/metadata/anvil/__init__.py
--- a/src/azul/plugins/metadata/anvil/__init__.py   (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/src/azul/plugins/metadata/anvil/__init__.py   (date 1756832717225)
@@ -444,6 +444,11 @@
                 field_schemas.append(self._pfb_schema_from_anvil_column(table_name=table_name,
                                                                         column_name='drs_uri',
                                                                         anvil_datatype='string'))
+            if table_name == 'anvil_dataset':
+                field_schemas.append(self._pfb_schema_from_anvil_column(table_name=table_name,
+                                                                        column_name='source_datarepo_snapshot_id',
+                                                                        anvil_datatype='string',
+                                                                        is_optional=False))
             for column_schema in table_schema['columns']:
                 field_schemas.append(
                     self._pfb_schema_from_anvil_column(table_name=table_name,
Index: test/service/data/manifest/verbatim/jsonl/anvil/linked.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/service/data/manifest/verbatim/jsonl/anvil/linked.json b/test/service/data/manifest/verbatim/jsonl/anvil/linked.json
--- a/test/service/data/manifest/verbatim/jsonl/anvil/linked.json   (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/service/data/manifest/verbatim/jsonl/anvil/linked.json   (date 1756832717254)
@@ -45,6 +45,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS",
             "version": "2022-06-01T00:00:00.000000Z"
         }
Index: test/indexer/data/2370f948-2783-aeb6-afea-e022897f4dcf.tdr.anvil.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/2370f948-2783-aeb6-afea-e022897f4dcf.tdr.anvil.json b/test/indexer/data/2370f948-2783-aeb6-afea-e022897f4dcf.tdr.anvil.json
--- a/test/indexer/data/2370f948-2783-aeb6-afea-e022897f4dcf.tdr.anvil.json (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/2370f948-2783-aeb6-afea-e022897f4dcf.tdr.anvil.json (date 1756832717248)
@@ -29,6 +29,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS",
             "version": "2022-06-01T00:00:00.000000Z"
         }
Index: test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json
--- a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json (date 1756832717246)
@@ -41,6 +41,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS",
             "version": "2022-06-01T00:00:00.000000Z"
         },
Index: test/indexer/data/c2711e94-9966-a0ef-88be-88caf3e8a29b.tdr.anvil.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/c2711e94-9966-a0ef-88be-88caf3e8a29b.tdr.anvil.json b/test/indexer/data/c2711e94-9966-a0ef-88be-88caf3e8a29b.tdr.anvil.json
--- a/test/indexer/data/c2711e94-9966-a0ef-88be-88caf3e8a29b.tdr.anvil.json (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/c2711e94-9966-a0ef-88be-88caf3e8a29b.tdr.anvil.json (date 1756832717249)
@@ -20,6 +20,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS",
             "version": "2022-06-01T00:00:00.000000Z"
         },
Index: test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json
--- a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json   (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json   (date 1756832717244)
@@ -1465,7 +1465,7 @@
     },
     {
         "_index": "azul_v2_nadove4_test_replica",
-        "_id": "anvil_dataset_2370f948-2783-4eb6-afea-e022897f4dcf_f10ae38d485330bf8fc7a95f7eb06626b984ad6a",
+        "_id": "anvil_dataset_2370f948-2783-4eb6-afea-e022897f4dcf_08dc3055f8c16284f1758f8ce7138f835118ddd7",
         "_score": 1.0,
         "_source": {
             "entity_id": "2370f948-2783-4eb6-afea-e022897f4dcf",
@@ -1497,6 +1497,7 @@
                 "source_datarepo_row_ids": [
                     "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
                 ],
+                "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
                 "title": "ANVIL_CMG_UWASH_DS_BDIS",
                 "version": "2022-06-01T00:00:00.000000Z"
             }
Index: test/indexer/data/6b35f59c-d33d-abf7-9ba0-c7b3a0ca82f3.tdr.anvil.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/6b35f59c-d33d-abf7-9ba0-c7b3a0ca82f3.tdr.anvil.json b/test/indexer/data/6b35f59c-d33d-abf7-9ba0-c7b3a0ca82f3.tdr.anvil.json
--- a/test/indexer/data/6b35f59c-d33d-abf7-9ba0-c7b3a0ca82f3.tdr.anvil.json (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/6b35f59c-d33d-abf7-9ba0-c7b3a0ca82f3.tdr.anvil.json (date 1756832717235)
@@ -37,6 +37,7 @@
             "source_datarepo_row_ids": [
                 "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
             ],
+            "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "title": "ANVIL_CMG_UWASH_DS_BDIS",
             "version": "2022-06-01T00:00:00.000000Z"
         }
Index: test/indexer/data/6c87f0e1-509d-46a4-b845-7584df39263b.tables.tdr.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/data/6c87f0e1-509d-46a4-b845-7584df39263b.tables.tdr.json b/test/indexer/data/6c87f0e1-509d-46a4-b845-7584df39263b.tables.tdr.json
--- a/test/indexer/data/6c87f0e1-509d-46a4-b845-7584df39263b.tables.tdr.json    (revision 7516806bf10000dc44c6cc7d75c4a6beb515adf6)
+++ b/test/indexer/data/6c87f0e1-509d-46a4-b845-7584df39263b.tables.tdr.json    (date 1756832717238)
@@ -88,6 +88,7 @@
                     "source_datarepo_row_ids": [
                         "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
                     ],
+                    "source_datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
                     "title": "ANVIL_CMG_UWASH_DS_BDIS"
                 }
             ]
```

--- a/src/azul/service/avro_pfb.py
+++ b/src/azul/service/avro_pfb.py
@@ -485,7 +485,7 @@ def avro_pfb_schema(azul_avro_schema: Iterable[JSON]) -> JSON:
 def _inject_reference_handover_columns(field_types: FieldTypes) -> FieldTypes:
     return {
         entity_type: (
-            dict(fields, datarepo_row_id=null_str, source_datarepo_snapshot_id=null_str)
+            dict(fields, datarepo_row_id=null_str, datarepo_snapshot_id=null_str)
             if isinstance(fields, dict) and 'source_datarepo_row_ids' in fields
             else fields
         )
@@ -496,7 +496,7 @@ def _inject_reference_handover_columns(field_types: FieldTypes) -> FieldTypes:
 def _inject_reference_handover_values(entity: MutableJSON, doc: JSON):
     if 'source_datarepo_row_ids' in entity:
         entity['datarepo_row_id'] = entity['document_id']
-        entity['source_datarepo_snapshot_id'] = one(doc['sources'])['id']
+        entity['datarepo_snapshot_id'] = one(doc['sources'])['id']
 
 
 # FIXME: It's not obvious as to why these are union types. Explain or change.
@@ -570,7 +570,7 @@ def _entity_schema_recursive(field_types: FieldTypes,
                 'estimated_cell_count',
                 'total_estimated_cells',
                 'total_estimated_cells_redundant',
-                'source_datarepo_snapshot_id',
+                'datarepo_snapshot_id',
             )
             path_exceptions = (
                 ('projects', 'accessions'),

--- a/src/azul/service/manifest_service.py
+++ b/src/azul/service/manifest_service.py
@@ -3,6 +3,9 @@ from abc import (
     abstractmethod,
 )
 import base64
+from bisect import (
+    insort,
+)
 from collections.abc import (
     Iterable,
     Mapping,
@@ -2059,6 +2062,18 @@ class PFBVerbatimManifestGenerator(FileBasedManifestGenerator,
         replicas = list(self._list_replicas())
         plugin = self.metadata_plugin
         replica_schemas = plugin.verbatim_pfb_schema(replicas)
+        if config.is_anvil_enabled(self.catalog):
+            for replica in replicas:
+                if replica['replica_type'] == 'anvil_dataset':
+                    source_id = replica['source']['id']
+                    replica['contents']['datarepo_snapshot_id'] = source_id
+            for schema in replica_schemas:
+                if schema['name'] == 'anvil_dataset':
+                    field_schema = plugin._pfb_schema_from_anvil_column(table_name='anvil_dataset',
+                                                                        column_name='datarepo_snapshot_id',
+                                                                        anvil_datatype='string',
+                                                                        is_optional=False)
+                    insort(schema['fields'], field_schema, key=itemgetter('name'))
         # Ensure field order is consistent for unit tests
         replica_schemas.sort(key=itemgetter('name'))
         links = {

--- a/src/azul/service/manifest_service.py
+++ b/src/azul/service/manifest_service.py
@@ -2062,6 +2062,8 @@ class PFBVerbatimManifestGenerator(FileBasedManifestGenerator,
         replicas = list(self._list_replicas())
         plugin = self.metadata_plugin
         replica_schemas = plugin.verbatim_pfb_schema(replicas)
+        # FIXME: Move injection of snapshot ID field to metadata plugin
+        #        https://github.com/DataBiosphere/azul/issues/7411
         if config.is_anvil_enabled(self.catalog):
             for replica in replicas:
                 if replica['replica_type'] == 'anvil_dataset':

--- a/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json
+++ b/test/service/data/manifest/verbatim/pfb/anvil/pfb_entities.json
@@ -425,6 +425,7 @@
                 "DS-BDIS"
             ],
             "datarepo_row_id": "2370f948-2783-4eb6-afea-e022897f4dcf",
+            "datarepo_snapshot_id": "6c87f0e1-509d-46a4-b845-7584df39263b",
             "dataset_id": "52ee7665-7033-63f2-a8d9-ce8e32666739",
             "owner": [
                 "Debbie Nickerson"

--- a/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json
+++ b/test/service/data/manifest/verbatim/pfb/anvil/pfb_schema.json
@@ -533,6 +533,11 @@
                             "type": "string"
                         },
                         {
+                            "name": "datarepo_snapshot_id",
+                            "namespace": "anvil_dataset",
+                            "type": "string"
+                        },
+                        {
                             "name": "dataset_id",
                             "namespace": "anvil_dataset",
                             "type": "string"

